### ISSUE TITLE
emcc-props.json leads to wrong path for JsonToItemsTaskFactory.dll

### ISSUE THREAD

**Issue by GeorgeS2019** [dotnet/runtime#59619]
When building the project: [blazor native](https://github.com/mattleibow/SkiaSharpBlazorWebAssembly/issues/2)

### This error happens 
Error	MSB4175:
The task factory "JsonToItemsTaskFactory.JsonToItemsTaskFactory" could not be loaded from the assembly "C:\Program Files\dotnet\packs\Microsoft.NET.Runtime.MonoTargets.Sdk\6.0.0-rtm.21474.5\Sdk\..\tasks\net472\JsonToItemsTaskFactory.dll". Could not load file or assembly 'System.Text.Json, Version=5.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The system cannot find the file specified.

### net472\JsonToItemsTaskFactory.dll is in the directory
"C:\Program Files\dotnet\packs\Microsoft.NET.Runtime.MonoTargets.Sdk\6.0.0-rtm.21474.5\tasks\net472\JsonToItemsTaskFactory.dll"

instead of

"C:\Program Files\dotnet\packs\Microsoft.NET.Runtime.MonoTargets.Sdk\6.0.0-rtm.21474.5\Sdk\..\tasks\net472\JsonToItemsTaskFactory.dll"

## The report problem is in the WasmApp.Native.targets	line => 408
C:\Program Files\dotnet\packs\Microsoft.NET.Runtime.WebAssembly.Sdk\6.0.0-rtm.21474.5\Sdk\WasmApp.Native.targets	line => 408	

### WasmApp.Native.targets line 408
    <ReadEmccProps JsonFilePath="$(_WasmRuntimePackSrcDir)emcc-props.json">
      <Output TaskParameter="EmccProperties" ItemName="_EmccPropItems" />
    </ReadEmccProps>

### The content of emcc-props.json
```json
{
  "items": {
    "EmccProperties": [
      { "identity": "RuntimeEmccVersion",     "value": "2.0.23" },
      { "identity": "RuntimeEmccVersionRaw",  "value": "emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 2.0.23 (b15ca40ca02f469de3bf9a4de67a2655aede8974)" },
      { "identity": "RuntimeEmccVersionHash", "value": "b15ca40ca02f469de3bf9a4de67a2655aede8974" }
    ]
  }
}
```

### Configuration

6.0.0-rtm.21474.5
VS Enterprise 2022 preview (64bit) version 17.0.0 Preview 4.1

**Comment by mattleibow** [comment]
We are also getting this now after bumping the SDK: https://github.com/dotnet/maui/pull/2659

However, we are using the full msbuild to build the solution.

**Comment by GeorgeS2019** [comment]
Just for FYI, the [side effect ](https://github.com/dotnet/installer/issues/12142) relevant the experience described here.

**Comment by mattleibow** [comment]
I figured out the issue. It is because the public preview of VS/MSBuild has older dlls. It works locally for me because I am also running the main build of VS.

However, since we use the public previews of VS on CI, I created this script to update the dlls in place in a VS install. I believe this goes away once P5 of VS is released: https://github.com/dotnet/maui/blob/2c1b919cf5132262a2328adba59a67141d0f12d0/eng/scripts/patch-vs.ps1

**Comment by lambdageek** [comment]
tl;dr - @mattleibow is right this is gonna be fixed by VS2022 Preview 5.

---

Details:

* There are 2 versions of MSBuild: a modern one using .NET 6 libraries that runs when you use the `dotnet build` CLI; and a .NET Framework 4.x MSBuild that ships with Visual Studio
* When MSBuild loads tasks, they go into the same process and should also be either the net6 or net472 versions.
* So when we build the Microsoft.NET.Runtime.MonoTargets.Sdk nuget, we put two copies of the tasks in there.  Fore net6 and for net472
* System.Text.Json is part of the .NET 6 runtime, but on .NET Framework it's an extra assembly.  MSBuild doesn't know anything about nuget, so we bundle this assembly with the net472 task.
* Visual Studio MSBuild also bundles System.Text.Json now because it uses it internally for something.
* When we build Microsoft.NET.Runtime.MonoTargets.Sdk it includes 2 task assemblies: RuntimeConfigParser.dll and JsonToItemsTaskFactory.dll
* Both our assemblies depend on System.Text.Json - JsonToItemsTaskFactory explicitly, and RuntimeConfigParser.dll implicitly through its dependency on MSBuild.
* When we put the Microsoft.NET.Runtime.MonoTargets.Sdk.nupkg together we copy files from a build directory into a package directory
* RuntimeConfigParser implicitly depends on STJ 4.0.1 apparently; but JsonToItemsTaskFactory explicitly depends on 5.0.0.
* For whatever reason we copy the STJ (4.0.1) from the RuntimeConfigParser build dir into the package (probably it just happens to copy it second).
* When we install the runtime pack, we end up in a situation where JsonToItemsTaskFactory has a reference to STJ 5.0.0, but next to it on disk is STJ 4.0.1
* With VS 2022 Preview 4 this is a problem because MSBuild in that preview bundles STJ 4.0.1, too.  So there's nowhere to find STJ 5.0.0.  So the JsonToItemsTaskFactory assembly can't be loaded.
* But with VS 2022 Preview 5, MSBuild started to bundle STJ 5.0.0.  So when we try to load the JsonToItemsTaskFactory.dll it will resolve the reference for STJ 5.0.0 to the one that is in the MSBuild directory.  So the assembly is loaded and life goes on.

I discussed with @steveisok  and we think it's too risky to change how the runtime is packaging JsonToItemsTaskFactory.dll and RuntimeConfigParser.dll tasks for .NET 6 RTM.

For .NET 7+ we need to:
- (good) Ship the version of STJ that JsonToItemsTaskFactory actually needs.
- (Better) don't create a situation where RuntimeConfigParser and JsonToItemsTaskFactory need different assemblies.  Actually just package both tasks into a single assembly.  That will also speed up builds a little because there is fewer DLLs to load.
- (best) there's also something going on where we bundle a copy of `Microsoft.Build.dll` into the `net472` (but not the net6) tasks directory.  Stop doing that.

**Comment by lambdageek** [comment]
The runtime pack packaging of the tasks was improved in #59720
